File: prototypes/firefox/crawler.go
package main

import (
	"fmt"
	"log"

	"github.com/tebeka/selenium"
)

const (
	targetURL    = "https://recaptcha-demo.appspot.com/recaptcha-v3-request-scores.php"
	seleniumPort = 4444
)

func main() {
	log.Println("Start")

	selenium.SetDebug(true)
	cap := selenium.Capabilities{"browserName": "firefox"}

	// Connect to the WebDriver instance running in a docker container.
	wd, err := selenium.NewRemote(cap, fmt.Sprintf("http://127.0.0.1:%d/wd/hub", seleniumPort))
	if err != nil {
		log.Fatalln("Failed to connect to webDriver: ", err)
	}
	defer wd.Quit()

	// Navigate to the target URL.
	if err := wd.Get(targetURL); err != nil {
		log.Println("Failed to connect to target URL: ", err)
		return
	}

	// Get an HTML tag for the reCAPTCHA step description.
	elem, err := wd.FindElement(selenium.ByCSSSelector, ".step1")
	if err != nil {
		log.Println("Failed to find HTML element: ", err)
		return
	}

	elemText, err := elem.Text()
	if err != nil {
		log.Println("Failed to get tag text: ", err)
		return
	}
	log.Println("Elem text: ", elemText)

	log.Println("End")
}

/*
# Reference
- https://pkg.go.dev/github.com/tebeka/selenium#pkg-overview

# Line Count
- Total:      14
- Reused:     0
- Written:    14
- Referenced: 0
*/

File: prototypes/firefox/crawler.py
from time import sleep
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait as Wait

print("Start")

# Initialise web driver for Firefox
options = webdriver.FirefoxOptions()
driver = webdriver.Remote(
  command_executor="http://127.0.0.1:4444",
  options=options,
)

print("Send HTTP GET request")
driver.get("https://recaptcha-demo.appspot.com/recaptcha-v3-request-scores.php")

print("Wait for 2 sec")
sleep(2)

# Get button to trigger reCAPTCHA action
print("Click button")
buttonElem = driver.find_element(By.CSS_SELECTOR, "button.go")
buttonElem.click()

print("Wait for 5 sec")
sleep(5)
responseElem = driver.find_element(By.CSS_SELECTOR, "pre.response")
print("Response element is displayed?: " + str(responseElem.is_displayed()))
print("Response text: " + responseElem.text)

driver.quit()
print("End")

"""
# REFERENCES
- https://www.selenium.dev/documentation/webdriver/remote_webdriver/
- https://selenium-python.readthedocs.io/getting-started.html
- https://stackoverflow.com/questions/15937966/in-python-selenium-how-does-one-find-the-visibility-of-an-element

# LINE COUNT
Total:
Reused:
Written:
Referenced: 3
"""
